### PR TITLE
Lfn changes

### PR DESCRIPTION
**Pull request by rob-c**
This PR contains the code to address #1040 . I've been nudged by several people over this for a few years and it's something I would like to see us have a better handle over.

It would be nice to move the code dealing with output file naming out of mass storage file but this is a serious undertaking to do correctly as it will require expanding the API as well as developing the test suite to make sure that this all works correctly.

With that in mind we currently also have #1055 as an open issue and this goes at least part of the way to solving the need to fix this more quickly.

I think a more complete version of this PR may be to clean up the fact this was cherry picked from #917 as well as adding in a solution for #1055 before looking into how to merge functionality between MassStorage and Dirac file objects.

### CODE REVIEW

**Comment by egede** [comment]
retest please

**Comment by rob-c** [comment]
@egede Something is up with the build system I assumed someone was working on it.
This hasn't been tested yet but should pass all non DIRAC/LHCb tests as it doesn't touch anything anywhere else. I'll give this another quick look when I get a chance, hopefully before the end of the week, I suspect there is a minor bug in GaudiExec from this branch tough.

**Comment by egede** [comment]
retest this please

**Comment by mesmith75** [comment]
@rob-c I think this is ok. Are you sure you switched all the RTHandlers? Would it be worth putting in a default job number just in case?

**Comment by rob-c** [comment]
@mesmith75 There is a bug here that will cause all GaudiExec jobs to not submit correctly to DIRAC. I'm fixing it atm.

**Comment by rob-c** [comment]
@mesmith75 This goes some way to fixing things.

`DiracFile.remoteDir` is now respected as the path extension to use when placing user-requested/managed LFN on the grid.

I've migrated the functionality of using `{jid}`, `{sjid}` and `{fname}` out from MassStorageFile.
This now means that `DiracFile.remoteDir` supports these constructs. (although I've not fully tested this yet)

Aka users can now do:

```
...
j.outputfile=[DiracFile(remoteDir='MyImportantWork_{jid}/{sjid}'...)]
...
```

This would then lead to an LFNs similar to:

```
/lhcb/user/r/rcurrie/MyImportantWork_4/0/2017_07/171565/171565261/tmpYEqjgq.txt
/lhcb/user/r/rcurrie/MyImportantWork_4/1/2017_07/171565/171565332/tmpYEqjgq.txt
```


There is an automatic naming option which can be used by default to place LFN in a path which is constructed using the jid and sjid (so gives similar to MassStorageFile defaults).
This has to be turned on through the configuration system as it can potentially lead to confusions.

This works as originally described in #1040 and has the advantage that this requires a 1 line change in the config to get the effects.


This now needs testing to submit real DiracFiles so I'll do this soon.

**Comment by egede** [comment]
retest this please

**Comment by egede** [comment]
retest this please

**Comment by alexanderrichards** [comment]
Can one of the admins verify this patch?

**Comment by mesmith75** [comment]
Yes let's put it into the bugfix release.

**Comment by rob-c** [comment]
👍 Righto, Do I have enough time to write a docstring example or 2 for this? merge by 2pm or so?

**Comment by rob-c** [comment]
Edit I'll get the changes in by 3pm, need to sort a few things out today,

**Comment by rob-c** [comment]
Ok, I'm curious to hear any feedback on the docstr for DiracFile I've tried to cover a few different use-cases rather than just outputfiles.

**Comment by mesmith75** [comment]
@rob-c They look good.

**Comment by rob-c** [comment]
@mesmith75 OK if there's nothing really to add I'm happy to merge this

**Comment by mesmith75** [comment]
Actually can I request a slight change to No 5 in the docstring? It just needs to be clear that the filename has to correspond to the file created by the job (DaVinci etc) in order for it to be uploaded and you can subsequently only change the base directory path.

**Comment by mesmith75** [comment]
@rob-c Thanks. The `remoteDir` has become a little confusing now but maybe it is just me.

To be clear it only affects the location of files uploaded at the completion of a job? It doesn't affect the location of file uploaded with `DiracFile.put()` ?

**Comment by rob-c** [comment]
@mesmith ah... That's the 1 use case I haven't explored. It obviously won't work if you have a file without a job parent and a jid in the name... I can have a look at this tomorrow of needed but it should work fine if you just change the remotedir before a put i think.
I can do some checks tomorrow now if needed.

There is a bit too much bluring of functionality here but a rewrite to handle mass storage and dirac in a more consistent way is difficult :/

**Comment by mesmith75** [comment]
@rob-c I don't think that is an issue. You can already specify the LFN in `put()` but the meaning of `remoteDir` is perhaps a little opaque.

**Comment by mesmith75** [comment]
retest this please

**Comment by rob-c** [comment]
@mesmith75 Think I discovered a minor bug in this PR, I'm checking now

**Comment by rob-c** [comment]
OK, I need to clean up this PR (not sure what has happened)

**Comment by rob-c** [comment]
Not sure what happened there, possibly the result of cherry-picking. The minor bug I spotted was that the put and WN script behaviour was completely inconsistent with the DIRAC upload at the end of the job. These should now be in sync.

**Comment by rob-c** [comment]
I've just run a test job and the LFN renaming appears to be working for me as well as the new config to have ganga put LFN in a different sub-folder

File: python/Ganga/test/GPI/FileTests/TestMassStorageClient.py [hunk 1]
```diff
@@ -113,6 +113,7 @@ def test_b_testClientSideComplete(self):
             # Check that the files were placed in the correct place on storage
             output_dir = os.path.join(self.outputFilePath, str(j.id), str(sj.id))
             for file_ in j.inputfiles:
+                print(os.path.join(output_dir, file_.namePattern))
```
Comment (mesmith75) [inline python/Ganga/test/GPI/FileTests/TestMassStorageClient.py]: Is that print supposed to remain?

**Comment by mesmith75** [comment]
@rob-c Thanks that looks great.

**Pull request merged**